cas-server-core-authentication/src/main/java/org/jasig/cas/authentication/AbstractAuthenticationManager.java

```diff
@@ -0,0 +1,317 @@
+package org.jasig.cas.authentication;
+
+import com.codahale.metrics.annotation.Counted;
+import com.codahale.metrics.annotation.Metered;
+import com.codahale.metrics.annotation.Timed;
+import org.jasig.cas.authentication.handler.support.HttpBasedServiceCredentialsAuthenticationHandler;
+import org.jasig.cas.authentication.principal.NullPrincipal;
+import org.jasig.cas.authentication.principal.Principal;
+import org.jasig.cas.authentication.principal.PrincipalResolver;
+import org.jasig.cas.authentication.principal.Service;
+import org.jasig.cas.services.RegisteredService;
+import org.jasig.cas.services.ReloadableServicesManager;
+import org.jasig.cas.services.UnauthorizedSsoServiceException;
+import org.jasig.inspektr.audit.annotation.Audit;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.util.Assert;
+
+import javax.annotation.Resource;
+import javax.validation.constraints.NotNull;
+import java.security.GeneralSecurityException;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.Iterator;
+import java.util.LinkedHashMap;
+import java.util.LinkedHashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * This is {@link AbstractAuthenticationManager}, which provides common operations
+ * around an authentication manager implementation.
+ *
+ * @author Misagh Moayyed
+ * @since 4.3.0
+ */
+public abstract class AbstractAuthenticationManager implements AuthenticationManager {
+    /** Log instance for logging events, errors, warnings, etc. */
+    protected final Logger logger = LoggerFactory.getLogger(getClass());
+
+    /**
+     * The Services manager.
+     */
+    protected ReloadableServicesManager servicesManager;
+
+
+    /** An array of AuthenticationAttributesPopulators. */
+    @NotNull
+    protected List<AuthenticationMetaDataPopulator> authenticationMetaDataPopulators =
+            new ArrayList<>();
+
+    /** Authentication security policy. */
+    @NotNull
+    protected AuthenticationPolicy authenticationPolicy = new AnyAuthenticationPolicy();
```

Comment:
Shouldn't the `AuthenticationPolicy` be part of the `PolicyBasedAuthenticationManager`?